daemon/logger/journald/journald.go

```diff
@@ -104,19 +148,110 @@ func validateLogOpt(cfg map[string]string) error {
 }
 
 func (s *journald) Log(msg *logger.Message) error {
-	vars := map[string]string{}
-	for k, v := range s.vars {
-		vars[k] = v
-	}
-	if msg.Partial {
-		vars["CONTAINER_PARTIAL_MESSAGE"] = "true"
-	}
-	if msg.Source == "stderr" {
-		return journal.Send(string(msg.Line), journal.PriErr, vars)
+	// build the message struct for the helper, and send it on down
+	message := MessageWithVars{
+		Message: *msg,
+		Vars:    s.vars,
 	}
-	return journal.Send(string(msg.Line), journal.PriInfo, vars)
+	s.writing.Lock()
+	defer s.writing.Unlock()
+	return s.encoder.Encode(&message)
 }
 
 func (s *journald) Name() string {
 	return name
 }
+
+func (s *journald) closeWriter() {
+	s.pipe.Close()
+	if err := s.cmd.Wait(); err != nil {
+		eerr, ok := err.(*exec.ExitError)
+		if !ok {
+			logrus.Errorf("error waiting on log handler: %v", err)
+			return
+		}
+		status, ok := eerr.Sys().(syscall.WaitStatus)
+		if !ok {
+			logrus.Errorf("error waiting on log handler: %v", err)
+			return
+		}
+		if !status.Signaled() || (status.Signal() != syscall.SIGTERM && status.Signal() != syscall.SIGKILL) {
+			logrus.Errorf("error waiting on log handler: %v", err)
+			return
+		}
+	}
+}
+
+func loggerLog(f string, args ...interface{}) {
+	s := fmt.Sprintf(f, args...)
+	journal.Send(s, journal.PriInfo, nil)
+	fmt.Fprintln(os.Stderr, s)
+}
+
+func joinScope(scope string) error {
+	// This is... not ideal.  But if we're here, we're just going to have
+	// to assume that we know how to compute the same path that runc is
+	// going to use, based on a value of the form "parent:docker:ID", where
+	// the "docker" is literal.
+	parts := strings.Split(scope, ":")
```

Comment:
Doing the derivation right using libcontainer makes sense, since it's already a dependency.  I'll have another look at that.  Though that does leave open a question - how often do we see a running journald when systemd isn't managing cgroups?